test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTests.cs

```diff
@@ -0,0 +1,103 @@
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.TestHost;
+using MvcTagHelpersWebSite;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public class MvcTagHelpersTests
+    {
+        private readonly IServiceProvider _provider = TestHelper.CreateServices("MvcTagHelpersWebSite");
+        private readonly Action<IApplicationBuilder> _app = new Startup().Configure;
+        private static readonly Assembly _resourcesAssembly = typeof(TagHelpersTests).GetTypeInfo().Assembly;
+
+        [Theory]
+        [InlineData("Index", null)]
+        [InlineData("Order", "/Order/Submit")]
+        [InlineData("Product", null)]
+        [InlineData("ProductSubmit", null)]
+        [InlineData("Customer", "/Customer/Customer")]
+        public async Task MvcTagHelpers_GeneratesExpectedResults(string action, string antiForgeryPath)
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+            var expectedMediaType = MediaTypeHeaderValue.Parse("text/html; charset=utf-8");
+
+            // The K runtime compiles every file under compiler/resources as a resource at runtime with the same name
+            // as the file name, in order to update a baseline you just need to change the file in that folder.
+            var expectedContent =
+                    await _resourcesAssembly.ReadResourceAsStringAsync
+                                     ("compiler/resources/MvcTagHelpersWebSite.MvcTagHelperTest." + action + ".html");
+
+            // Act
+            // The host is not important as everything runs in memory and tests are isolated from each other.
+            var response = await client.GetAsync("http://localhost/MvcTagHelperTest/" + action);
+            var responseContent = await response.Content.ReadAsStringAsync();
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Assert.Equal(expectedMediaType, response.Content.Headers.ContentType);
+
+            if (antiForgeryPath != null)
+            {
+                var forgeryToken = AntiForgeryTestHelper.RetrieveAntiForgeryToken(responseContent, antiForgeryPath);
+                expectedContent = string.Format(expectedContent, forgeryToken);
+            }
+            Assert.Equal(expectedContent.Trim(), responseContent.Trim());
+        }
+
+        [Fact]
+        public async Task ValidationTagHelpers_GeneratesExpectedSpansAndDivs()
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+            var response = await client.GetAsync("http://localhost/MvcTagHelperTest/Customer");
```

Comment:
looks very strange to GET in the `// Arrange` section.  instead, follow the
``` c#
  // Arrange - 1
  // Act - 1
  // Assert - 1

  // Arrange - 2
  // Act - 2
  // Assert - 2
```
pattern. use about the same assertions for `// Assert - 1` as in the above `[Theory]`.  then remove the "Customer" case from the above `[Theory]` (since it would be redundant).

another option might be to disable the anti-forgery token generation (`<form asp-anti-forgery="false">`) and just do the POST in this test method.